### PR TITLE
Pricing RFC

### PR DESCRIPTION
**Pull request by Pikkupomo**

### CODE REVIEW

File: shoop/core/models/product_shops.py [hunk 1]
```diff
@@ -56,6 +56,8 @@ class ShopProduct(models.Model):
         "ProductMedia", null=True, blank=True, related_name="primary_image_for_shop_products", on_delete=models.SET_NULL
     )
 
+    default_price = MoneyField(verbose_name=_("Default price"))
```
Comment (akx) [inline shoop/core/models/product_shops.py]: Would be nice to document the semantics of this field. (How is its taxfulness determined?)

File: shoop/core/models/products.py [hunk 2]
```diff
@@ -258,20 +258,28 @@ def get_cheapest_child_price(self, context, quantity=1):
             for c in self.variation_children.all()
         )[0]
 
-    def get_price(self, context, quantity=1):
+    def get_price_info(self, context, quantity=1):
         """
         :type context: shoop.core.contexts.PriceTaxContext
-        :rtype: shoop.core.pricing.Price
+        :rtype: PriceInfo
```
Comment (akx) [inline shoop/core/models/products.py]: Qualify the rtype.

File: shoop/core/models/products.py [hunk 3]
```diff
@@ -258,20 +258,28 @@ def get_cheapest_child_price(self, context, quantity=1):
...
         module = get_pricing_module()
         pricing_context = module.get_context(context)
-        return module.get_price(pricing_context, product_id=self.pk, quantity=quantity)
+        return module.get_price_info(pricing_context, product=self, quantity=quantity)
 
-    def get_base_price(self):
+    def get_price(self, context, quantity=1):
+        """
+        :type context: shoop.core.contexts.PriceTaxContext
+        :rtype: Price
```
Comment (akx) [inline shoop/core/models/products.py]: Qualify the rtype.

File: shoop/core/models/products.py [hunk 4]
```diff
@@ -258,20 +258,28 @@ def get_cheapest_child_price(self, context, quantity=1):
...
+        """
+        return self.get_price_info(context, quantity).price
+
+    def get_base_price(self, context):
```
Comment (akx) [inline shoop/core/models/products.py]: So what does `get_base_price` do anyway?

File: shoop/core/pricing/__init__.py [hunk 5]
```diff
@@ -116,26 +117,26 @@ def get_context_from_request(self, request):
     def get_context_from_data(self, **context_data):
         return self.pricing_context_class(**context_data)
 
-    def get_price(self, context, product_id, quantity=1):
+    def get_price_info(self, context, product, quantity=1):
+        return PriceInfo(TaxlessPrice(0), TaxlessPrice(0))
```
Comment (akx) [inline shoop/core/pricing/__init__.py]: Use kwargs? Would be more explicit (and I CBA to scroll around to find the signature of `PriceInfo`)

File: shoop/core/pricing/__init__.py [hunk 6]
```diff
@@ -116,26 +117,26 @@ def get_context_from_request(self, request):
...
+
+    def get_price(self, context, product, quantity=1):
         """
         Get context-specific Price for the given product and quantity.
 
-        May return TaxlessPrice or TaxfulPrice.
```
Comment (akx) [inline shoop/core/pricing/__init__.py]: Why was this removed?

File: shoop/core/pricing/__init__.py [hunk 7]
```diff
@@ -116,26 +117,26 @@ def get_context_from_request(self, request):
...
         """
-        Get base Price for the given product.
-
-        May return TaxlessPrice or TaxfulPrice.
+        Get context-specific base Price for the given product.
 
+        :type context: PricingContext
         :rtype: Price
         """
-        return TaxlessPrice(0)
```
Comment (akx) [inline shoop/core/pricing/__init__.py]: Same here for kwargs.

File: shoop/core/pricing/__init__.py [hunk 8]
```diff
@@ -155,16 +156,16 @@ def get_pricing_steps(self, context, product_id):
         """
         return [(0, TaxlessPrice(0))]
 
-    def get_prices(self, context, product_ids, quantity=1):
```
Comment (akx) [inline shoop/core/pricing/__init__.py]: The point in `product_ids` vs `product`s was that I foresaw many situations when you didn't need the full product objects, just the IDs.

File: shoop/simple_pricing/admin_form_part.py [hunk 9]
```diff
@@ -26,7 +26,7 @@ def __init__(self, **kwargs):
             self._build_fields()
 
     def _build_fields(self):
-        self.shops = [None] + list(Shop.objects.all())
+        self.shops = list(Shop.objects.all())
```
Comment (akx) [inline shoop/simple_pricing/admin_form_part.py]: We can't have shopless simple prices anymore?

File: shoop/simple_pricing/module.py [hunk 10]
```diff
@@ -60,19 +58,41 @@ def get_price(self, context, product_id, quantity=1):
...
+
+        price_cls = TaxfulPrice if includes_tax else TaxlessPrice
+        return PriceInfo(price_cls(price), price_cls(base_price))
+
+    def get_base_price(self, context, product):
+        """
+        Find a base price for product in given shop.
+        Only finds base price, doesn't care about customer groups and such.
+
+        :param context:
```
Comment (akx) [inline shoop/simple_pricing/module.py]: Either fill these in or get rid of them.

File: shoop/simple_pricing/module.py [hunk 11]
```diff
@@ -60,19 +58,41 @@ def get_price(self, context, product_id, quantity=1):
...
+
+        base_price = price
+
+        price_cls = TaxfulPrice if includes_tax else TaxlessPrice
+        return PriceInfo(price_cls(price), price_cls(base_price))
+
+    def get_base_price(self, context, product):
+        """
+        Find a base price for product in given shop.
+        Only finds base price, doesn't care about customer groups and such.
```
Comment (akx) [inline shoop/simple_pricing/module.py]: Newline required.

File: shoop/testing/factories.py [hunk 12]
```diff
@@ -548,8 +554,11 @@ def create_random_order(customer=None, products=(), completion_probability=0):
     if not customer:
         raise ValueError("No valid contacts")
 
+    shop = Shop.objects.first()
```
Comment (akx) [inline shoop/testing/factories.py]: This should probably use `get_default_shop()`

File: shoop/core/pricing/__init__.py (hunk 6 above)
Comment (Pikkupomo) [inline shoop/core/pricing/__init__.py]: At some point of my iterations this returned `PriceInfo` perhaps a leftover from there :)

File: shoop/testing/factories.py [hunk 13]
```diff
@@ -434,9 +434,15 @@ def create_order_with_product(product, supplier, quantity, taxless_unit_price, t
...
     order.full_clean()
     order.save()
+
+    request = RequestFactory().get("/")
+    request.shop = get_default_shop()
+
     for x in range(n_lines):
         product_order_line = OrderLine(order=order)
-        update_order_line_from_product(request=None, order_line=product_order_line, product=product, quantity=quantity,
+        update_order_line_from_product(request=request,
```
Comment (akx) [inline shoop/testing/factories.py]: Maybe this API should require a `PriceTaxContext` instead of a `request`...

File: shoop_tests/core/test_order_creator.py [hunk 14]
```diff
@@ -57,7 +58,10 @@ def test_order_creator(admin_user):
         require_verification=True,
     ))
 
-    creator = OrderCreator(request=None)
+    request = RequestFactory().get("/")
+    request.shop = Shop.objects.first()
```
Comment (akx) [inline shoop_tests/core/test_order_creator.py]: `get_default_shop()` here too, methinks.

File: shoop/core/pricing/__init__.py (hunk 8 above)
Comment (Pikkupomo) [inline shoop/core/pricing/__init__.py]: I understand that, the problem is that we need `ShopProduct` (atleast in simple pricing) which means we must query it from there every time. Thought this could be a wise direction in that sense.

File: shoop_tests/core/test_price_info.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
+# -*- coding: utf-8 -*-
+# This file is part of Shoop.
+#
+# Copyright (c) 2012-2015, Shoop Ltd. All rights reserved.
+#
+# This source code is licensed under the AGPLv3 license found in the
+# LICENSE file in the root directory of this source tree.
+from _pytest.python import raises
```
Comment (akx) [inline shoop_tests/core/test_price_info.py]: Unused import, and also a private import?

File: shoop/simple_pricing/admin_form_part.py (hunk 9 above)
Comment (Pikkupomo) [inline shoop/simple_pricing/admin_form_part.py]: No

File: shoop/testing/factories.py (hunk 12 above)
Comment (Pikkupomo) [inline shoop/testing/factories.py]: agreed!

**Comment by akx** [comment]
Hmmm.. Re the test changes touching `OrderCreator()`; the `request` parameter there is marked

```python
self.request = request  # TODO: Get rid of `request`?
```

anyway, and it's not used in too many places in there anyway, so...

Maybe add a second optional parameter `pricing_context` or something that could be used in the `update_order_line_from_product()` call(s) in the OrderCreator.

If a `request` is given and the context isn't, the context should naturally be automatically derived from the request in the ctor.

File: shoop/simple_pricing/admin_form_part.py (hunk 9 above)
Comment (Pikkupomo) [inline shoop/simple_pricing/admin_form_part.py]: Edit 2: And `Product` has no price cached, everything comes from pricing module which determines that, so perhaps it might be possible anyways

File: shoop/core/pricing/__init__.py (hunk 8 above)
Comment (akx) [inline shoop/core/pricing/__init__.py]: You can get a ShopProduct via `ShopProduct.objects.get(shop_id=, product_id=)`, can't you? :)

File: shoop/simple_pricing/module.py [hunk 16]
```diff
@@ -60,19 +62,41 @@ def get_price(self, context, product_id, quantity=1):
...
-        return _create_price(price, includes_tax)
-
-    def get_base_price(self, product):
+        if result:
+            (price, includes_tax) = result[0]
+            if default_price > 0:
+                price = min([default_price, price])
+        else:
+            price = default_price
+            includes_tax = False
```
Comment (akx) [inline shoop/simple_pricing/module.py]: So yeah, this is suspect. You get a `default_price` which is a Decimal, then just assume it's taxless...

File: shoop/core/pricing/__init__.py (hunk 8 above)
Comment (Pikkupomo) [inline shoop/core/pricing/__init__.py]: ...and cause a query to database? :) this is what I was trying to avoid :) Not sure if that big of a problem now, but when querying 2 million products, maybe then?

File: shoop/testing/factories.py (hunk 13 above)
Comment (Pikkupomo) [inline shoop/testing/factories.py]: For what reason?

File: shoop/core/pricing/__init__.py [hunk 17]
```diff
@@ -125,17 +132,21 @@ def get_price(self, context, product_id, quantity=1):
...
-        Get base Price for the given product.
+        Get context-specific base Price for the given product.
 
         May return TaxlessPrice or TaxfulPrice.
 
+        :type context: PricingContext
         :rtype: Price
         """
-        return TaxlessPrice(0)
+        return PriceInfo(
```
Comment (akx) [inline shoop/core/pricing/__init__.py]: So wouldn't it be easier to just return a `TaxlessPrice(0)` now... :D

File: shoop/core/pricing/__init__.py (hunk 8 above)
Comment (akx) [inline shoop/core/pricing/__init__.py]: Then passing a list of 2 million products to the function is also a problem.

`get_shop_instance()` _does_ use the Product instance's built-in cache of `ShopProduct`s (or hits the database when it can't), that's true, but you could allow your lowest-level API to work with just a `product_id` and a `shop_product` (that a slightly-higher level API of your own passes in).

Comment (akx) [inline shoop/core/pricing/__init__.py]: Also, do note that if you make this change, we'll be forced to bump the version to 2.0, as this is a backward-incompatible API change. :)

File: shoop/core/pricing/price_info.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+
+class PriceInfo(object):
+    def __init__(self, price, base_price):
+        assert isinstance(price, Price)
+        assert isinstance(base_price, Price)
+        assert price._units_match(base_price)
+        self.price = price
+        self.base_price = base_price
+        self.includes_tax = bool(isinstance(price, TaxfulPrice))
+        self.expires_on = None
```
Comment (akx) [inline shoop/core/pricing/price_info.py]: Should this maybe be an optional argument for the ctor?

File: shoop/core/pricing/price_info.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+# LICENSE file in the root directory of this source tree.
+
+from decimal import Decimal
+
+from shoop.core.pricing import Price, TaxfulPrice
+
+
+class PriceInfo(object):
+    def __init__(self, price, base_price):
+        assert isinstance(price, Price)
```
Comment (akx) [inline shoop/core/pricing/price_info.py]: A docstring here might be muy bueno.

File: shoop/core/settings.py [hunk 20]
```diff
@@ -136,3 +136,6 @@
 
 #: The submission URL for Shoop's telemetry (statistics) system
 SHOOP_TELEMETRY_URL = "https://telemetry.shoop.io/collect/"
+
+#: Defines if Shoop handles prices taxful by default for `ShopProduct.default_price`
```
Comment (akx) [inline shoop/core/settings.py]: Maybe

> Defines whether the default prices for products are assumed to be taxful or taxless.

File: shoop/testing/factories.py (hunk 13 above)
Comment (akx) [inline shoop/testing/factories.py]: Because this is the only thing it does with the `request`:

```python
context = PriceTaxContext.from_request(request)
```

Comment (akx) [inline shoop/testing/factories.py]: And that'd make it more HTTP-agnostic anyway.

File: shoop_tests/core/test_order_creator.py [hunk 21]
```diff
@@ -101,7 +109,10 @@ def test_order_source_parentage(admin_user):
         unit_price=TaxlessPrice(5),
         parent_line_id="parent"
     ))
-    creator = OrderCreator(request=None)
+    request = RequestFactory().get("/")
+    request.shop = Shop.objects.first()
```
Comment (akx) [inline shoop_tests/core/test_order_creator.py]: `get_default_shop()`?

File: shoop_tests/core/test_price_info.py [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+#
+# This source code is licensed under the AGPLv3 license found in the
+# LICENSE file in the root directory of this source tree.
+
+from shoop.core.pricing import TaxlessPrice
+from shoop.core.pricing.price_info import PriceInfo
+
+
+def test_price_info_no_discount():
+    pi = PriceInfo(TaxlessPrice(100), TaxlessPrice(100))
```
Comment (akx) [inline shoop_tests/core/test_price_info.py]: Named parameters?

File: shoop_tests/core/test_price_info.py [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+    assert (pi.is_discounted == False)
```
Comment (akx) [inline shoop_tests/core/test_price_info.py]: No need for parentheses, and `assert not pi.is_discounted` sounds more fluent anyhoo.

File: shoop/core/pricing/__init__.py [hunk 24]
```diff
@@ -116,7 +117,13 @@ def get_context_from_request(self, request):
     def get_context_from_data(self, **context_data):
         return self.pricing_context_class(**context_data)
 
-    def get_price(self, context, product_id, quantity=1):
+    def get_price_info(self, context, product, quantity=1):
+        return PriceInfo(
+            price=TaxlessPrice(0),
+            base_price=TaxlessPrice(0)
+        )
```
Comment (suutari-ai) [inline shoop/core/pricing/__init__.py]: Do we really want to implement it here? IMHO this base class should be just an interface (i.e. abstract). Probably best to declare it as ABCMeta and decorate get_price_info with `@abstractmethod`.

File: shoop/core/pricing/__init__.py [hunk 25]
```diff
@@ -125,17 +132,21 @@ def get_price(self, context, product_id, quantity=1):
         :type context: PricingContext
         :rtype: Price
         """
-        return TaxlessPrice(0)
+        return self.get_price_info(context, product, quantity).price
```
Comment (suutari-ai) [inline shoop/core/pricing/__init__.py]: `get_price` method should be removed from this interface, since it's not really needed and if someone happens to redefine it so it no longer returns `get_price_info(...).price` then it will make just confusion. Product has get_price helper and that is fine, but at this level it's more harm than helpful.

File: shoop/core/pricing/__init__.py [hunk 26]
```diff
@@ -125,17 +132,21 @@ def get_price(self, context, product_id, quantity=1):
...
         May return TaxlessPrice or TaxfulPrice.
 
+        :type context: PricingContext
         :rtype: Price
         """
-        return TaxlessPrice(0)
+        return PriceInfo(
+            price=TaxlessPrice(0),
+            base_price=TaxlessPrice(0)
+        ).base_price
```
Comment (suutari-ai) [inline shoop/core/pricing/__init__.py]: Same for `get_base_price` (as for `get_price`): It should be removed from `PricingModule`.

Comment (suutari-ai) [inline shoop/core/pricing/__init__.py]: Also this definition is wrong anyway (shouldn't return hardcoded value, but `get_price_info(...).base_price`, but I'd rather see this whole method being removed).

File: shoop/core/pricing/__init__.py (hunk 8 above)
Comment (suutari-ai) [inline shoop/core/pricing/__init__.py]: I thought we almost decided to bump to 2.0 anyway?

That said, I'd rather use just "products" in the API and make it accept both Product objects and ids of such. (Pre-condition being: `isinstance(product, Product) or (isinstance(product, int) and Product.object.filter(pk=product).exists())`).

File: shoop/core/pricing/price_info.py (hunk 18 above)
Comment (suutari-ai) [inline shoop/core/pricing/price_info.py]: It should.

File: shoop/core/pricing/price_info.py [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+        assert isinstance(base_price, Price)
+        assert price._units_match(base_price)
+        self.price = price
+        self.base_price = base_price
+        self.includes_tax = bool(isinstance(price, TaxfulPrice))
```
Comment (suutari-ai) [inline shoop/core/pricing/price_info.py]: Don't do isinstance, but rather use `price.includes_tax`

Comment (suutari-ai) [inline shoop/core/pricing/price_info.py]: Though, includes_tax should be a read-only property implemented with a getter rather than pure data member. To (1) make it stand out that it is a derived property, (2) make it harder to do stupid things like `pi = PriceInfo(TaxfulPrice(...)); pi.includes_tax = False`

File: shoop/core/pricing/price_info.py [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+        self.expires_on = None
+
+    def get_discount_amount(self):
+        return (self.base_price - self.price)
```
Comment (suutari-ai) [inline shoop/core/pricing/price_info.py]: Make this `@property discount_amount` to make it easier to access (it's computationally cheap too).

File: shoop/core/pricing/price_info.py [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+
+    def get_discount_percentage(self):
+        return Decimal((1 - self.price / self.base_price) * 100)
```
Comment (suutari-ai) [inline shoop/core/pricing/price_info.py]: `(self.price / self.base_price)` is already decimal, you don't need the cast, i.e. make it just:
```python
return (1 - (self.price / self.base_price)) * 100
```

Comment (suutari-ai) [inline shoop/core/pricing/price_info.py]: And this should be `@property` too.

File: shoop/core/settings.py (hunk 20 above)
Comment (suutari-ai) [inline shoop/core/settings.py]: How about Methods and coupons etc.?

I don't like this variable, since it will drive us away from the goal that each Shop could define its prices with or without tax even if another Shop in the same installation wants to do the opposite.

File: shoop/front/templates/shoop/front/macros.jinja [hunk 30]
```diff
@@ -13,7 +14,7 @@
         <div class="caption">
             <h4 class="name">{{ product.name }}</h4>
             {# TODO: (TAX) Check following line (gets taxful price of product)  #}
-            <p class="price">{{ product.get_price(request)|home_currency }}</p>
+            <p class="price">{{ price_info.price|home_currency }} {% if price_info.is_discounted %}({{price_info.base_price|home_currency}}) <span class="discount-percent">{{ -price_info.get_discount_percentage()|int }}%</span>{% endif %}</p>
```
Comment (suutari-ai) [inline shoop/front/templates/shoop/front/macros.jinja]: Split this to several lines to make it easier to review.

Does the `|int` jinja-filter convert the value to int by flooring or by rounding?

File: shoop/front/templates/shoop/front/macros.jinja [hunk 31]
```diff
@@ -33,7 +34,8 @@
                                 <span class="image">{{ product.primary_image|thumbnail(size=(220, 230))|default(STATIC_URL + "site/img/no_image.png") }}</span>
                                 <span class="name">{{ product.name }}</span>
                                 {# TODO: (TAX) Check following line (gets taxful price of product)  #}
-                                <span class="price inverse big">{{ product.get_price(request)|home_currency }}</span>
+                                {% set price_info = product.get_price_info(request) %}
+                                <span class="price inverse big">{{ price_info.base_price|home_currency }}</span>
```
Comment (suutari-ai) [inline shoop/front/templates/shoop/front/macros.jinja]: Why are you converting from price to base_price here? (Price would be correct imho)

File: shoop/simple_pricing/module.py [hunk 32]
```diff
@@ -60,19 +61,30 @@ def get_price(self, context, product_id, quantity=1):
...
+        )
 
+    def get_base_price(self, context, product):
+        product_price_model = SimpleProductPrice.objects.filter(product=product, group=None).first()
+        if product_price_model:
+            includes_tax = product_price_model.includes_tax
+            base_price = product_price_model.price
+        else:
+            base_price = product.get_shop_instance(context.shop_id).default_price
+            includes_tax = settings.SHOOP_DEFAULT_PRICES_INCLUDE_TAX
```
Comment (suutari-ai) [inline shoop/simple_pricing/module.py]: Why are you ignoring the customer group here? Didn't we decide that with SimplePricing price=base_price always?  ... Anyway.. this method should probably not exists (see comments above).

File: shoop/testing/factories.py (hunk 13 above)
Comment (suutari-ai) [inline shoop/testing/factories.py]: I agree that it should take a PriceTaxContext rather than request.

File: shoop_tests/core/test_order_creator.py [hunk 33]
```diff
@@ -57,7 +58,10 @@ def test_order_creator(admin_user):
         require_verification=True,
     ))
 
-    creator = OrderCreator(request=None)
+    request = RequestFactory().get("/")
+    request.shop = get_default_shop()
+
+    creator = OrderCreator(request)
```
Comment (suutari-ai) [inline shoop_tests/core/test_order_creator.py]: By the same logic as above.. Shouldn't OrderCreator also take a PriceTaxContext rather than a request? Though I don't know what it does with the request anyway...

Comment (suutari-ai) [inline shoop_tests/core/test_order_creator.py]: OK, seems that I'm not the first that thought it. OrderCreator class currently starts with:
```python
class OrderCreator(object):
    def _init__(self, request):
        self.request = request  # TODO: Get rid of `request`?
...
```

Comment (suutari-ai) [inline shoop_tests/core/test_order_creator.py]: Seems that OrderCreator uses request to update_order_line_from_product call and for getting information about the order creating initiator: IP address and logged-in user account. I also passes the request to the order_creator_finished signal. Dunno if the last one is a good idea though... (maybe not?)

File: shoop_tests/core/test_price_info.py (hunk 22 above)
Comment (suutari-ai) [inline shoop_tests/core/test_price_info.py]: pi is fine here, even p would do.

Comment (suutari-ai) [inline shoop_tests/core/test_price_info.py]: Or π if you want to piss python 2 devs :) (not really)

File: shoop_tests/core/test_price_info.py [hunk 34]
```diff
@@ -0,0 +1,24 @@
...
+    assert (pi.get_discount_percentage() == 0)
+    assert (pi.get_discount_amount() == TaxlessPrice(0))
+
+
+def test_price_info_discounts():
+    pi = PriceInfo(TaxlessPrice(50), TaxlessPrice(100))
+    assert (pi.is_discounted == True)
+    assert (pi.get_discount_percentage() == 50)
+    assert (pi.get_discount_amount() == TaxlessPrice(50))
```
Comment (suutari-ai) [inline shoop_tests/core/test_price_info.py]: 50 % is the worst test case for testing discount percentages, since then it wouldn't catch if its calculated correctly as 1-x/y rather than the incorrect x/y. How about p = PriceInfo(TaxlessPrice(75), TaxlessPrice(100) instead?.

File: shoop_tests/simple_pricing/test_form.py [hunk 35]
```diff
@@ -16,7 +17,10 @@
 
 def _get_test_product():
     shop = get_default_shop()
-    product = create_product("Just-A-Pricing-Product")
+    product = create_product("Just-A-Pricing-Product", shop)
+    sp, created = ShopProduct.objects.get_or_create(product=product, shop=shop)
+    sp.default_price = 200
+    sp.save()
     SimpleProductPrice.objects.create(product=product, shop=None, price=200, includes_tax=False)
```
Comment (suutari-ai) [inline shoop_tests/simple_pricing/test_form.py]: This shop=None line should probably be removed. Though it should not pass the tests, since shop should be required field for SimpleProductPrice new? Shouldn't it?

File: shoop/core/models/product_shops.py [hunk 36]
```diff
@@ -56,6 +56,10 @@ class ShopProduct(models.Model):
         "ProductMedia", null=True, blank=True, related_name="primary_image_for_shop_products", on_delete=models.SET_NULL
     )
 
+    # the default price of this product in the shop, taxfulness is determined in
+    # `settings.SHOOP_DEFAULT_PRICES_INCLUDE_TAX`
+    default_price = MoneyField(verbose_name=_("Default price"))
```
Comment (suutari-ai) [inline shoop/core/models/product_shops.py]: Taxfulness should be determined by self.shop.prices_include_tax instead...

File: shoop/core/pricing/price_info.py (hunk 19 above)
Comment (suutari-ai) [inline shoop/core/pricing/price_info.py]: For the PriceInfo class at least.

File: shoop_tests/core/test_price_info.py (hunk 22 above)
Comment (suutari-ai) [inline shoop_tests/core/test_price_info.py]: Ah.. you mean like PriceInfo(price=somethig, base_price=something_else)?

File: shoop_tests/simple_pricing/test_form.py (hunk 35 above)
Comment (Pikkupomo) [inline shoop_tests/simple_pricing/test_form.py]: @akx says shop is still required. I removed it once and discussed about this yesterday with him, and im still not 100% sure which is the correct way to go.
